frontend/class-twitter.php

```diff
@@ -9,364 +9,532 @@
+		if ( preg_match( '`([A-Za-z0-9_]{1,25})$`', $id, $match ) ) {
+			return $match[1];
+		} else {
+			return false;
 		}
+	}
 
+	/**
+	 * Displays the domain tag for the site.
+	 */
+	protected function site_domain() {
 		/**
-		 * Displays the URL for Twitter.
+		 * Filter: 'wpseo_twitter_domain' - Allow changing the Twitter domain as output in the Twitter card by WP SEO
 		 *
-		 * Only used when OpenGraph is inactive.
+		 * @api string $unsigned Name string
 		 */
-		public function twitter_url() {
-			/**
-			 * Filter: 'wpseo_twitter_url' - Allow changing the URL as output in the Twitter card by WP SEO
-			 *
-			 * @api string $unsigned Canonical URL
-			 */
-			$url = apply_filters( 'wpseo_twitter_url', $this->canonical( false ) );
-			if ( is_string( $url ) && $url !== '' ) {
-				$this->output_metatag( 'url', esc_url( $url ), true );
-			}
+		$domain = apply_filters( 'wpseo_twitter_domain', get_bloginfo( 'name' ) );
+		if ( is_string( $domain ) && $domain !== '' ) {
+			$this->output_metatag( 'domain', $domain );
 		}
+	}
 
-		/**
-		 * Outputs a Twitter image tag for a given image
-		 *
-		 * @param string $img
-		 * @return bool
-		 */
-		public function image_output( $img ) {
+	/**
+	 * Displays the image for Twitter
+	 *
+	 * Only used when OpenGraph is inactive or Summary Large Image card is chosen.
+	 */
+	protected function image() {
 
-			/**
-			 * Filter: 'wpseo_twitter_image' - Allow changing the Twitter Card image
-			 *
-			 * @api string $img Image URL string
-			 */
-			$img = apply_filters( 'wpseo_twitter_image', $img );
+		if ( 'gallery' === $this->type ) {
+			$this->gallery_images_output();
+		} else {
+			$this->single_image_output();
+		}
+
+		if ( count( $this->shown_images ) == 0 && $this->options['og_default_image'] !== '' ) {
+			$this->image_output( $this->options['og_default_image'] );
+		}
+	}
 
-			$escaped_img = esc_url( $img );
+	/**
+	 * Outputs the first 4 images of a gallery as the posts gallery images
+	 */
+	private function gallery_images_output() {
+		$images = get_post_gallery_images();
+
+		// If there are no images attached, use the standard single image output
+		if ( count( $images ) === 0 ) {
+			$this->single_image_output();
+
+			return;
+		}
+
+		$image_counter = 0;
+		foreach ( $images as $image ) {
+			if ( $image_counter > 3 ) {
+				return;
+			}
+			$this->image_output( $image, 'image' . $image_counter );
+			$image_counter ++;
+		}
+	}
 
-			if ( in_array( $escaped_img, $this->shown_images ) ) {
-				return false;
+	/**
+	 * Takes care of image output when we only need to display a single image.
+	 */
+	private function single_image_output() {
+		if ( is_singular() ) {
+			if ( $this->homepage_image_output() ) {
+				return;
+			}
+			if ( $this->image_from_meta_values_output() ) {
+				return;
+			}
+			if ( $this->image_thumbnail_output() ) {
+				return;
+			}
+			if ( $this->image_from_content_output() ) {
+				return;
+			}
+		} else {
+			if ( $this->homepage_image_output() ) {
```

Comment:
This piece is executed anyway, maybe it could be placed above this if-else statement